api/src/main/java/org/pmiops/workbench/cohortbuilder/util/QueryParameterValues.java

```diff
@@ -65,35 +103,83 @@ public static String replaceNamedParameters(QueryJobConfiguration queryJobConfig
     return Matchers.replaceAllInMap(patternToReplacement, result);
   }
 
-  public static String formatQuery(String query) {
+  @NotNull
+  public static String formatQuery(@NotNull String query) {
     return new BasicFormatterImpl().format(query);
   }
 
   // use lookbehind for non-word character, since "'"(@" or " @" don't represent left-side word
   // boundaries.
-  private static Pattern buildParameterRegex(String parameterName) {
+  @NotNull
+  private static Pattern buildParameterRegex(@NotNull String parameterName) {
     return Pattern.compile(String.format("(?<=\\W)%s\\b", decorateParameterName(parameterName)));
   }
 
-  public static String decorateParameterName(String parameterName) {
+  @NotNull
+  public static String decorateParameterName(@NotNull String parameterName) {
     return "@" + parameterName;
   }
 
-  private static String getReplacementString(QueryParameterValue parameterValue) {
+  @Nullable
+  public static QueryParameterValue toTimestampQpv(@Nullable OffsetDateTime offsetDateTime) {
+    final String arg =
+        Optional.ofNullable(offsetDateTime).map(QPV_TIMESTAMP_FORMATTER::format).orElse(null);
+    return QueryParameterValue.timestamp(arg);
+  }
+
+  // Return null instead of Optional.empty() so the return value can go directly into
+  // the content map of an InsertAllRequest.RowToInsert.
+  @Nullable
+  public static String toInsertRowString(@Nullable OffsetDateTime offsetDateTime) {
+    return Optional.ofNullable(offsetDateTime)
+        .map(ROW_TO_INSERT_TIMESTAMP_FORMATTER::format)
+        .orElse(null);
+  }
+
+  // BigQuery TIMESTAMP types don't include a zone or offset, but are always UTC.
+  public static Optional<OffsetDateTime> rowToInsertStringToOffsetTimestamp(
+      @Nullable String bqTimeString) {
+    return Optional.ofNullable(bqTimeString)
+        .filter(s -> s.length() > 0)
+        .map(ROW_TO_INSERT_TIMESTAMP_FORMATTER::parse)
+        .map(LocalDateTime::from)
+        .map(ldt -> OffsetDateTime.of(ldt, ZoneOffset.UTC));
+  }
+
+  @NotNull
```

Comment:
The other way to operate would have been to convert to string when writing the DTO, but I felt like that was losing type safety too early in the process.